pin-project-internal/src/pin_project/derive.rs

```diff
@@ -311,54 +362,98 @@ impl<'a> Context<'a> {
-        let (where_clause_fields, where_clause_ref_fields) = match fields {
-            Fields::Named(_) => {
-                (quote!(#where_clause #proj_fields), quote!(#where_clause #proj_ref_fields))
-            }
-            Fields::Unnamed(_) => {
-                (quote!(#proj_fields #where_clause;), quote!(#proj_ref_fields #where_clause;))
-            }
+        let (where_clause_fields, where_clause_ref_fields, where_clause_own_fields) = match fields {
+            Fields::Named(_) => (
+                quote!(#where_clause #proj_fields),
+                quote!(#where_clause #proj_ref_fields),
+                quote!(#orig_where_clause #proj_own_fields),
+            ),
+            Fields::Unnamed(_) => (
+                quote!(#proj_fields #where_clause;),
+                quote!(#proj_ref_fields #where_clause;),
+                quote!(#proj_own_fields #orig_where_clause;),
+            ),
             Fields::Unit => unreachable!(),
         };
 
-        let proj_items = quote! {
+        let mut proj_items = quote! {
             #[allow(clippy::mut_mut)] // This lint warns `&mut &mut <ty>`.
             #[allow(dead_code)] // This lint warns unused fields/variants.
             #vis struct #proj_ident #proj_generics #where_clause_fields
             #[allow(dead_code)] // This lint warns unused fields/variants.
             #vis struct #proj_ref_ident #proj_generics #where_clause_ref_fields
         };
 
-        let proj_body = quote! {
+        if let Some(replace) = self.replace {
+            proj_items.extend(quote_spanned! { replace =>
+                #[allow(dead_code)] // This lint warns unused fields/variants.
+                #vis struct #proj_own_ident #orig_generics #where_clause_own_fields
+            });
+        }
+
+        let proj_mut_body = quote! {
             let #orig_ident #proj_pat = self.get_unchecked_mut();
-            #proj_ident #proj_init
+            #proj_ident #proj_body
         };
         let proj_ref_body = quote! {
             let #orig_ident #proj_pat = self.get_ref();
-            #proj_ref_ident #proj_init
+            #proj_ref_ident #proj_body
         };
-        let proj_impl = self.make_proj_impl(&proj_body, &proj_ref_body);
+        let proj_own_body = quote! {
+            let __self_ptr: *mut Self = self.get_unchecked_mut();
+            let #orig_ident #proj_pat = &mut *__self_ptr;
+
+            // First, extract all the unpinned fields
+            let __result = #proj_own_ident #proj_move;
+
+            // Destructors will run in reverse order, so next create a guard to overwrite
+            // `self` with the replacement value without calling destructors.
+            let __guard = ::pin_project::#private::UnsafeOverwriteGuard {
+                target: __self_ptr,
+                value: ::core::mem::ManuallyDrop::new(__replacement),
+            };
+
+            // Now create guards to drop all the pinned fields
+            #proj_drop
```

Comment:
IIUC, `let __guard = ..` and `#proj_drop` need to wrap with a block. If not, ProjectionOwned is not dropped when a pinned field is panicked: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=7a49a30c2ecb55e722af08da23201465

Reply:
This... seems like a compiler bug. I created a reduced example: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=5e68a9ecbb2f4d61fff4faef592c2f07

I'm pretty sure the compiler is *supposed* to call the destructor in this situation.

Reply:
Oh, this certainly seems to be a compiler bug: https://github.com/rust-lang/rust/issues/47949